Add SubsetConstraint

ngfreiter opened the pull request:
This adds support for processing `SubsetConstraint`s and adds tests of this processing.
This PR is 4/7. These PRs impact:
* `shr-grammar` https://github.com/standardhealth/shr-grammar/pull/46
* `shr-text-import` https://github.com/standardhealth/shr-text-import/pull/108
* `shr-models` https://github.com/standardhealth/shr-models/pull/47
* `shr-expand`
* `shr-fhir-export` https://github.com/standardhealth/shr-fhir-export/pull/161
* `shr-json-schema-export` https://github.com/standardhealth/shr-json-schema-export/pull/16
* `shr-json-javadoc` https://github.com/standardhealth/shr-json-javadoc/pull/29

cmoesel commented on lib/expand.js:
```diff
@@ -695,6 +697,47 @@ class Expander {
+        const optionIdentifiers = targetVal.aggregateOptions.map(opt => opt.identifier);
+        const areValidOptions = constraint.subsetList.every(elem => {
+          return optionIdentifiers.some(optId => optId.equals(elem));
+        });
```
I see code below (line 727) to support child types, but the code above (line 709) seems like it won't work unless every subset item is _exactly_ one of the choice options.  E.g., if the original choice is `Animal or Plant`, then subset `Dog or Cat or MulberryBush` would cause an error since none of those `equals` `Animal` or `Plant` (even though they are based on `Animal`).

cmoesel replied:
Of course, this only comes out if `onValue` is set -- but I think that will be the case for most of these since they deal with choices.

cmoesel commented on lib/expand.js:
```diff
@@ -695,6 +697,47 @@ class Expander {
+        continue;
+      }
+      // Check if the each element of the new subset has some base type in the previous subset
+      // Does allow elements in previous subset to serve as base type for multiple elements in new subset
```
I don't see a test for this, and based on what I see in lines 707 - 716, I kind of doubt that child types work.  Do they?

cmoesel commented on lib/expand.js:
```diff
@@ -695,6 +697,47 @@ class Expander {
 
+  consolidateSubsetConstraint(element, value, constraint, previousConstraints) {
+    constraint = constraint.clone();
+    // If the constraint is actually on the target's value, update the path to explicitly include the target's value
```
Note that the code below doesn't actually do what the comment says (update the path).  Since we're dealing with choices, I don't think it *can* update the path unless it put a keyword `Value` in there (but I don't remember if we support that anyway).

ngfreiter commented:
That's a good point. The way it is right now, it looks like the constraint types have to explicitly be part of the original subset. But then if it is re-constrained, those new constraints can be children of the old subset constraints. Which is probably confusing. I can change one or the other, but do we have a preference between supporting child types or not?

cmoesel commented:
I think the preference is to support child types unless it introduces significant complexity.

ngfreiter commented:
I don't think it should be more complex, I believe just using `checkHasBaseType` instead of checking for equality should do the trick.

ngfreiter commented on lib/expand.js:
```diff
@@ -695,6 +697,47 @@ class Expander {
+        return previous.subsetList.some(tPrevious => this.checkHasBaseType(tCurrent, tPrevious));   
+      });
+      if (!subsetContained) {
+        logger.error('Cannot further constrain type of %s from %s to %s. ERROR_CODE:12015', target.toString(), previous.subsetList, constraint.subsetList);
```
Hey Nick, remember that logging framework you helped implement, consider using it!

cmoesel replied:
Ha.  I guess I missed that one too!

ngfreiter commented:
Fixed the logs, and changed the subset rules to be consistent. Now child types can be used. The changes are here https://github.com/standardhealth/shr-expand/pull/48/commits/9d821add17d6c0f21ea900af2879bf1099f72d8a.
